drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,869 @@
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	} else if (func_is_active_fe[id](dma)) {
+		LOG_ERR("FiFo error.");
+		dma_stm32_dump_stream_irq(dev, id);
+		dma_stm32_clear_stream_irq(dev, id);
+
+		stream->dma_callback(stream->callback_arg, id, -EIO);
+#endif
+	} else {
+		LOG_ERR("Transfer Error.");
+		dma_stm32_dump_stream_irq(dev, id);
+		dma_stm32_clear_stream_irq(dev, id);
+
+		stream->dma_callback(stream->callback_arg, id, -EIO);
+	}
+}
+
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+static void dma_stm32_config_direct_mode(DMA_TypeDef *dma, u32_t id, bool on)
+{
+	if (on) {
+		LL_DMA_EnableIT_DME(dma, table_stream[id]);
+		LL_DMA_DisableIT_FE(dma, table_stream[id]);
+		LL_DMA_DisableFifoMode(dma, table_stream[id]);
+	} else {
+		LL_DMA_DisableIT_DME(dma, table_stream[id]);
+		LL_DMA_EnableIT_FE(dma, table_stream[id]);
+		LL_DMA_EnableFifoMode(dma, table_stream[id]);
+	}
+}
+#endif
+
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+#define BURST_SIZE(dev, dev_short)					\
+do {									\
+	case 1:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_SINGLE);	\
+		break;							\
+	case 4:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC4);	\
+		break;							\
+	case 8:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC8);	\
+		break;							\
+	case 16:							\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC16);	\
+		break;							\
+	default:							\
+		LOG_ERR("Burst size error");				\
+		return -ENOTSUP;					\
+} while (0)
+#endif
+
+static u32_t dma_stm32_width_config(struct dma_config *config,
+				    bool source_periph,
+				    DMA_TypeDef *dma,
+				    u32_t id)
+{
+	u32_t periph, memory;
+	u32_t m_size = 0, p_size = 0;
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	u32_t periph_burst, memory_burst;
+#endif
+
+	if (source_periph) {
+		periph = config->source_data_size;
+		memory = config->dest_data_size;
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+		periph_burst = config->source_burst_length;
+		memory_burst = config->dest_burst_length;
+#endif
+	} else {
+		periph = config->dest_data_size;
+		memory = config->source_data_size;
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+		periph_burst = config->dest_burst_length;
+		memory_burst = config->source_burst_length;
+#endif
+	}
+	int index = find_lsb_set(config->source_data_size) - 1;
+
+	m_size = table_m_size[index];
+	index = find_lsb_set(config->dest_data_size) - 1;
+	p_size = table_p_size[index];
```

Comment:
Here m_size and p_size are wrong. For example in my case (I2S) I passed 1 for bot source and dest data size, because 
it is 16-bit transfer. But this code returned m_size = 0 and p_size = 0, which is wrong.
I remove the '- 1' to have index = 1 and it worked, but I'm not sure it is always correct.
Take a closer look here...

Reply:
I'm a bit confused about this, m_size and p_size come from source_data_size and dest_data_size, which are width of source data/dest data (in bytes) as introduced in include/dma.h:121. So if we pass 1 for source_data_size and dest_data_size then m_size and p_size just should be 0 for representing one byte, right? In your case it is 16-bits, so we should pass 2 for 2 bytes to source_data_size/dest_data_size,  right?

Reply:
i2s_ll_stm32.c:847 do gives source_data_size and dest_data_size 1 and commented them as '16 bit', but I believe this is the problem, they should be both assigned with 2, along with code in i2s_ll_stm32.c:865, do you think so?

Reply:
I believe you are right. It seems that you have to specify the number of bytes (i.e 1 is 8-bits,
2 is 16-bits, 4 is 32-bits, 8 is 64-bits). I saw that all other DMAs drivers are doing that way.

I don't know why, but seems that before it was using dma_width_index() and in that case the
mapping was different (i.e. 0 was 8-bits, 1 was 16-bits, 2 was 32-bits and so on). But I tend
to think that this routine is not matching the API expectation.

So, yes, I think that we need to change i2s_ll_stm32.c driver to match source_data_size and 
dest_data_size (i.e. put 2 there to mean 16-bit).